naga/src/proc/overloads/mod.rs

```diff
@@ -0,0 +1,325 @@
+use alloc::vec::Vec;
+use core::fmt;
+
+#[allow(rustdoc::private_intra_doc_links)]
+/// A trait for types representing of a set of Naga IR type rules.
+///
+/// Given an expression like `max(x, y)`, there are multiple type rules that
+/// could apply, depending on the types of `x` and `y`, like:
+///
+/// - `max(i32, i32) -> i32`
+/// - `max(vec4<f32>, vec4<f32>) -> vec4<f32>`
+///
+/// and so on. Borrowing WGSL's terminology, Naga calls the full set of type
+/// rules that might apply to a given expression its "overload candidates", or
+/// "overloads" for short.
+///
+/// This trait is meant to be implemented by types that represent a set of
+/// overload candidates. For example, [`MathFunction::overloads`] returns an
+/// `impl OverloadSet` describing the overloads for the given Naga IR math
+/// function. Naga's typifier, validator, and WGSL front end use this trait for
+/// their work.
+///
+/// [`MathFunction::overloads`]: ir::MathFunction::overloads
+///
+/// # Automatic conversions
+///
+/// In principle, overload sets are easy: you simply list all the overloads the
+/// function supports, and then when you're presented with a call to typecheck,
+/// you just see if the actual argument types presented in the source code match
+/// some overload from the list.
+///
+/// However, Naga supports languages like WGSL, which apply certain [automatic
+/// conversions] if necessary to make a call fit some overload's requirements.
+/// This means that the set of calls that are effectively allowed by a given set
+/// of overloads can be quite large, since any combination of automatic
+/// conversions might be applied.
+///
+/// For example, if `x` is a `u32`, and `100` is an abstract integer, then even
+/// though `max` has no overload like `max(u32, AbstractInt) -> ...`, the
+/// expression `max(x, 100)` is still allowed, because AbstractInt automatically
+/// converts to `u32`.
+///
+/// [automatic conversions]: https://gpuweb.github.io/gpuweb/wgsl/#feasible-automatic-conversion
+///
+/// # How to use `OverloadSet`
+///
+/// The general process of using an `OverloadSet` is as follows:
+///
+/// - Obtain an `OverloadSet` for a given operation (say, by calling
+///   [`MathFunction::overloads`]). This set is fixed by Naga IR's semantics.
+///
+/// - Call its [`arg`] method, supplying the type of the argument passed to the
+///   operation at a certain index. This returns a new `OverloadSet` containing
+///   only those overloads that could accept the given type for the given
+///   argument. This includes overloads that only match if automatic conversions
+///   are applied.
+///
+/// - If, at some point, the overload set becomes empty, then the set of
+///   arguments is not allowed for this operation, and the program is invalid.
+///   The `OverloadSet` trait provides an [`is_empty`] method.
+///
+/// - After all arguments have been supplied, if the overload set is still
+///   non-empty, you can call its `most_preferred` method to find out which
```

Comment:
nitpick: You've linked the other methods, so let's link this one, too!

```suggestion
///   non-empty, you can call its [`most_preferred`] method to find out which
```